extensions/metrics/src/__tests__/acceptance/metrics-push.acceptance.ts

```diff
@@ -0,0 +1,74 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/rest-explorer
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {RestApplication, RestServerConfig} from '@loopback/rest';
+import {givenHttpServerConfig, supertest} from '@loopback/testlab';
+import * as os from 'os';
+import pEvent from 'p-event';
+import * as path from 'path';
+import {promisify} from 'util';
+import {MetricsComponent} from '../..';
+import {MetricsBindings} from '../../keys';
+import {MetricsOptions} from '../../types';
+
+const runShell = require('@loopback/build/bin/utils').runShell;
+
+// Only run the test on Travis with Linux
+const verb =
+  process.env.TRAVIS && os.platform() === 'linux' ? describe : describe.skip;
+verb('Metrics (with push gateway)', function() {
+  // eslint-disable-next-line no-invalid-this
+  this.timeout(30000);
```

Comment:
Is this going to increase the duration of `npm test` by another 10-30 seconds? I am concerned that `npm test` is already taking to long to finish to make TDD practical, I am reluctant to make it even worse.

Reply:
So how long does it usually take to get `prom/pushgateway` docker container up and running? When I run `npm test` on my local machine for the second time, how long delay will be introduced because of waiting for docker?

> We may have to define a nightly build to run the costly integration tests. What do you think?

I feel it's not enough to run these tests nightly. If a pull request breaks one of these tests, then we will discover the problem too late.

Can we use the approach I have in place for running `repository-test` tests against real databases? Here is the gist:
- These tests ARE NOT run as part of `npm test`.
- There are clear instruction how to run the tests locally - see e.g. [MySQL instructions](https://github.com/strongloop/loopback-next/tree/7752b08664cd077272cada93266cd31a4c7fe199/acceptance/repository-mysql#running-the-test-suite).
-  The tests expect external services like databases to be already running and available. This way we pay the cost of starting the services only once, not for every test run.
- There is a single Travis CI job for each test suite (MongoDB, MySQL, etc.), these jobs are executed in parallel with other jobs like `npm test`, code linting, commit linting, etc. Example job config: [.travis.yml#L53-L69](https://github.com/strongloop/loopback-next/blob/7752b08664cd077272cada93266cd31a4c7fe199/.travis.yml#L53-L69) - it does not use Docker, because native MySQL is faster to setup, but Travis CI does support docker.

Reply:
Can you write a mock-up pushgateway that will call our metrics endpoint the same way as the real gateway does? That way we can verify push functionality from `extensions/metrics` tests.

Then we can add a new package to `acceptance` directory, e.g. `acceptance/push-metrics`, where we will use a real push gateway running in a docker container, to ensure our push implementation works with real gateways too and catch any discrepancies between our mock gateway and real gateways.

Reply:
Can we remove this 30 second timeout now, or at least reduce it to something like 5-10s?

Also I see that you introduce a mock push gateway, which is great! But how can we be sure that it's accurately simulating the behavior of a real gateway? Shouldn't we have an acceptance tests using the docker-based gateway as I proposed in my comment above?

Reply:
I think the mock-up push gateway gives us enough confidence as we just to have make sure this component is pushing metrics to the gateway (the correctness should have been covered by the prom-client).